atom-publisher-lib/src/main/scala/com/gu/atom/data/DynamoDataStore.scala

```diff
@@ -1,104 +1,160 @@
-      succeed(atom)
-    } catch {
-      case e: Exception => fail(handleException(e))
+  /**
+    * Conditional put, ensuring passed revision is higher than the value in dynamo
+    */
+  protected def put(json: Json, revision: Long): DataStoreResult[Json] = {
+    val expressionAttributeValues = new util.HashMap[String, Object]()
+    expressionAttributeValues.put(":revision", revision.asInstanceOf[Object])
+
+    Try {
+      table.putItem(
+        jsonToItem(json),
+        "contentChangeDetails.revision < :revision",
+        null,
+        expressionAttributeValues
+      )
+    } match {
+      case Success(_) => Right(json)
+      case Failure(conditionError: ConditionalCheckFailedException) =>
+        Left(VersionConflictError(revision))
+      case Failure(e) => Left(handleException(e))
     }
   }
 
-  private def handleException(e: Exception) = e match {
-    case serviceError: AmazonServiceException => DynamoError(serviceError.getErrorMessage)
-    case clientError: AmazonClientException => ClientError(clientError.getMessage)
-    case _ => ReadError
+  protected def delete(key: DynamoCompositeKey): DataStoreResult[DeleteItemResult] = {
+    Try {
+      key match {
+        case DynamoCompositeKey(partitionKey, None) =>
+          table.deleteItem(SimpleKeyName, partitionKey)
+        case DynamoCompositeKey(partitionKey, Some(sortKey)) =>
+          table.deleteItem(CompositeKey.partitionKey, partitionKey, CompositeKey.sortKey, sortKey)
+      }
+    } match {
+      case Success(outcome) => Right(outcome.getDeleteItemResult)
+      case Failure(e) => Left(handleException(e))
+    }
+  }
+
+  protected def scan: DataStoreResult[List[Json]] = {
+    Try {
+      table.scan().iterator.asScala.toList
+
+    } match {
+      case Success(items) => items.map(item => parseJson(item.toJSON)).sequenceU
+      case Failure(e) => Left(DynamoError(e.getMessage))
+    }
   }
 
   private def uniqueKey(dynamoCompositeKey: DynamoCompositeKey) = dynamoCompositeKey match {
-    case DynamoCompositeKey(partitionKey, None) => UniqueKey(KeyEquals('id, partitionKey))
-    case DynamoCompositeKey(partitionKey, Some(sortKey)) => UniqueKey(KeyEquals('atomType, partitionKey) and KeyEquals('id, sortKey))
+    case DynamoCompositeKey(partitionKey, None) =>
+      new PrimaryKey(SimpleKeyName, partitionKey)
+    case DynamoCompositeKey(partitionKey, Some(sortKey)) =>
+      new PrimaryKey(CompositeKey.partitionKey, partitionKey, CompositeKey.sortKey, sortKey)
+  }
+
+  def parseJson(s: String): DataStoreResult[Json] =
+    parser.parse(s).leftMap(parsingFailure => DynamoError(parsingFailure.getMessage))
+
+  def jsonToAtom(json: Json): DataStoreResult[Atom] =
+    json.as[Atom](backwardsCompatibleAtomDecoder).leftMap(error => DecoderError(error.message))
+
+  def jsonToItem(json: Json): Item = {
+    val item = new Item()
+    json.asObject.foreach { obj =>
+      obj.toMap.map { case (key, value) => item.withJSON(key, value.noSpaces) }
```

Comment:
Fortunately not. `withJson` requires an attribute name, which means you cannot just add the json object at the top level. So we have to insert each top level value individually, as json.